### ISSUE TITLE
db: recovery sync error edge cases

### ISSUE THREAD

**Issue by jbowens** [cockroachdb/pebble#984]
This facebook/rocksdb#7621 pull request prompted some investigation of recovery edge cases, in particular surrounding fsync errors. @petermattis and I talked a bit over slack and failed to see how the sequence of events described in facebook/rocksdb#7621 could occur. Specifically, the purported problematic case is:
1. a sync operation of the manifest begins, syncing a version edit _x_
2. the process crashes
3. the process recovers, reading the previous (pre-sync, without version edit _x_) state of the MANIFEST and then crashes
4. the process recovers again, now reading the new (post-sync, with version edit _x_) state of the MANIFEST

If this sequence of events is possible, step 3 can cause corruption by reusing file numbers that were already used by the unseen version edit _x_.

Typically, this case would not happen because the read of the MANIFEST in step 3 would still read the modified MANIFEST state containing version edit _x_ from the OS's buffer cache, even if the original sync operation from step 1 is still pending. Because RocksDB supports direct I/O, I think it makes sense that they don't get any such protection and need to guard against this possibility. In facebook/rocksdb#7621, Facebook asserts they witnessed this sequence of events also while running RocksDB with a remote, distributed filesystem. Is this a concern for any CockroachDB deployments?

The big question is do we have a guarantee across all supported filesystems that a previous process's MANIFEST write will never become visible _after_ we've already started recovery.

I remembered reading that the linux sync guarantees are very weak if there's ever an error syncing, and I did a little reading on "fsyncgate" [1][2][3]. Pebble already does the primary takeaway from fsyncgate, which is crash whenever there's an error during a MANIFEST write. However, I think that isn't the _only_ takeaway from fsyncgate, because a sync error can cause consistency issues after the process issuing the fsync syscall exits. If there's an error during a sync, some versions of Linux may mark the buffers as 'clean' even though they were never synced to disk:
> Linux < 4.13: fsync() errors can be lost in various ways; also buffers are marked clean after errors, so retrying fsync() can falsely report success and the modified buffer can be thrown away at any time due to memory pressure

If there's memory pressure and the modified buffer is thrown away, the MANIFEST may appear to go _back_ in time to the previous state of the MANIFEST, losing the version edit _x_ after recovery has already started. (This is a different sequence  than facebook/rocksdb#7621 addresses). I think this sequence is also problematic, because recovery might read a version edit _x_ that deletes a file _y_, finish recovery and physically delete the obsolete file _y_, crash and then recover from a MANIFEST that no longer contains version edit _x_ and does require the file _y_. One solution to this problem would be to rotate MANIFEST on recovery before deleting obsolete files. This ensures that the MANIFEST you recovered from is actually fully synced to disk because syncing it is a step in recovery. Currently, I believe we wait until the first new edit before we create a new MANIFEST. 

tldr:
1. Open question: do we have a guarantee across all CRDB-supported filesystems that a previous process's MANIFEST write will never become visible _after_ we've already started recovery?
2. It appears to be possible for a MANIFEST to lose previously-read version edits in a subsequent recovery, if at some point in the past there was an error syncing the MANIFEST.

@sumeerbhola and @itsbilal: do you have thoughts or see anything I'm missing here?

[1] https://danluu.com/fsyncgate/
[2] https://lwn.net/Articles/752063/
[3] https://wiki.postgresql.org/wiki/Fsync_Errors

**Comment by jbowens** [comment]
I think the second scenario also applies if there's a crash, recovery from unsynced MANIFEST edits in the OS page cache and then a hard reset. The writes removing files may sync while the MANIFEST edits marking those files as deleted may never have been synced.

**Comment by petermattis** [comment]
If I'm understanding the fix in facebook/rocksdb#7621 correctly, it seems like Pebble already has that protection:

>   // If the last sst file is not tracked in MANIFEST,
  // or the VersionEdit for the last sst file is not synced,
  // on recovery, the last sst file should be deleted,
  // and new sst files shouldn't reuse its file number.

During replay, Pebble never reuses a previously used sstable number:

```
		// Don't reuse any obsolete file numbers to avoid modifying an
		// ingested sstable's original external file.
		if d.mu.versions.nextFileNum <= fn {
			d.mu.versions.nextFileNum = fn + 1
		}
```

This bit of code was added in https://github.com/cockroachdb/pebble/pull/788 due to the way Pebble handles ingestion differently than RocksDB, but I think it has the same effect as the fix in facebook/rocksdb#7621.

> Open question: do we have a guarantee across all CRDB-supported filesystems that a previous process's MANIFEST write will never become visible after we've already started recovery?
>
> It appears to be possible for a MANIFEST to lose previously-read version edits in a subsequent recovery, if at some point in the past there was an error syncing the MANIFEST.

We always rotate the MANIFEST on recovery. Does that save us from these problematic scenarios? I haven't thought this through all the way. What I'm hoping is that the recovery is "atomic". We read some version of the MANIFEST, replay the WALS, flushing and creating new sstables, and only do a single `logAndApply` call to install the recovered state. If we crash during the recovery and restart and see more data in the MANIFEST we should be able to proceed with recovery again because all of the previous recovery work has been lost and because we never reuse file numbers we won't interact with any of that state.

Definitely worth more eyes and brains on this problem.
